src/modules/ekf2/EKF/python/ekf_derivation/derivation.py

```diff
@@ -205,6 +205,63 @@ def predict_covariance(
 
     return P_new
 
+def predict_vel_pos_closed_form(
+    state: VState,
+    d_vel: sf.V3,
+    d_vel_dt: sf.Scalar,
+    d_ang: sf.V3,
+    d_ang_dt: sf.Scalar,
+    g: sf.Scalar,
+    epsilon: sf.Scalar
+) -> (sf.V3, sf.V3):
+
+    # Closed-form integration of accelerometer and gyro measurements based on
+    # Goppert, James, et al. "A Closed-form Solution for the Strapdown Inertial Navigation Initial Value Problem." arXiv preprint arXiv:2310.04886 (2023).
+
+    # TODO: check which dt to use
+    state = vstate_to_state(state)
+    gyro = d_ang / d_ang_dt
+    accel = d_vel / d_vel_dt
+    dt = d_vel_dt
+    R_0 = state["quat_nominal"].to_rotation_matrix()
+    P_0 = sf.M32.block_matrix([[state["vel"], state["pos"]]])
+    R_l_prime = sf.Rot3.from_tangent(d_ang).to_rotation_matrix()
+    R_r_prime = sf.M33.eye()
+
+    A_M = sf.M32([
+        [0, 0],
+        [0, 0],
+        [g, 0]
+    ])
+    A_N = sf.M32.block_matrix([[accel, sf.V3()]])
+    B = sf.M22([
+        [0, 1],
+        [0, 0]
+    ])
+
+    def P(omega, A, B) -> sf.M32:
+        theta = (gyro.dot(gyro))**0.5
+        C1 = (1 - theta**2 / 2 - sf.cos(theta)) / sf.Max(theta**2, epsilon)
```

Comment:
It is better to  use switch statements with taylor series approximations near zero. In CogniPilot we are doing this here: https://github.com/CogniPilot/cyecca/blob/main/cyecca/symbolic.py#L346

Reply:
Good idea, thanks!